### DISCUSSION TITLE
target program not found

### DISCUSSION THREAD

**Discussion by wangzhankun** [xmake-io/xmake#2066] [Q&A]
我在桌面上使用xmake命令创建了一个工程。debug正常运行。开发两天后将工程移动到了其他目录，在vscode下使用debug按钮启动debug时提示，target program not found。我删除了build、.vscode、.xmake目录重新进行编译，也执行了`xmake c`，但是仍然提示错误。

**Comment by wangzhankun** [comment]
我发现了问题所在。当把工程转为vs之后，使用visual studio调试时，会弹出有个dll文件找不到。把dll文件复制到执行文件路径即可。请问xmake执行程序时为啥会看不到类似的提示呢？使用`xmake r`执行时会有如下报错：
`error: execv(E:\aboutme\研究生\科研活动\测绘项目\openssl_gm_client\build\windows\x86\release\openssl_gm_client.exe) failed(-1073741515)`
但是visual studio会有弹窗报错提示dll找不到。请问我该如何设置呢？

**Comment by waruqi** [comment]
> debug时提示，target program not found。

跟 dll 没关系，切到 vscode，需要点底部 build 按钮先编译一次，即使之前命令行编译过

>\openssl_gm_client\build\windows\x86\release\openssl_gm_client.exe) failed(-1073741515)
但是visual studio会有弹窗报错提示dll找不到。请问我该如何设置呢？

不知道，反正命令行该有的错误，vscode的终端输出也都会有，没有本质区别

**Reply by wangzhankun** [reply]
> 

确实跟dll没关系。我用build按钮试过了还是不行。run按钮工作正常。确认是debug模式下。

**Comment by wangzhankun** [comment]
* vscode版本号：1.64.2
* xmake版本号：2.6.3

**Reply by waruqi** [reply]
那你自己拉下 xmake-vscode 源码加载调下

**Comment by wangzhankun** [comment]
经过测试发现，是文件夹路径中包含中问的话，就会无法进行debug。切换到全英文的路径就可以了。
* Windows 11
* vscode 1.64.2
* xmake-vscode插件 1.4.9

**Reply by waruqi** [reply]
> 经过测试发现，是文件夹路径中包含中问的话，就会无法进行debug。切换到全英文的路径就可以了。
> 
> * Windows 11
> * vscode 1.64.2
> * xmake-vscode插件 1.4.9

你把这个文件拉下来，https://github.com/xmake-io/xmake-vscode/blob/master/assets/targetpath.lua

在自己项目执行下 `xmake l targetpath.lua` 看看，能否返回 target path

或者调下 下面这段逻辑，看看中文下，哪里break 了

https://github.com/xmake-io/xmake-vscode/blob/ec6667ad695e079febcb160f5bac607f04e131de/src/xmake.ts#L762-L770

**Reply by wangzhankun** [reply]
`xmake l targetpath.lua`可以成功找到路径，即使含有中文

**Reply by wangzhankun** [reply]
执行到765行之后，能够取出targetpath的路径，但是中文字符显示错误。
![image](https://user-images.githubusercontent.com/35724924/155835841-7c9d023b-fd28-4a86-88b2-2abe1260a0a1.png)

**Reply by waruqi** [reply]
> 执行到765行之后，能够取出targetpath的路径，但是中文字符显示错误。 ![image](https://user-images.githubusercontent.com/35724924/155835841-7c9d023b-fd28-4a86-88b2-2abe1260a0a1.png)

那就是取出来 编码不对了

**Comment by wangzhankun** [comment]
执行`xmake l targetpath.lua`显示输出的中文结果是正确的。
![image](https://user-images.githubusercontent.com/35724924/155835915-795df316-1b07-4d1a-b9f1-2d4dcc4759ef.png)

**Reply by waruqi** [reply]
目前是走 lua 执行后，print + iorun 捕获的输出，重定向输出，在windows上，默认作为 ansi 写入的，然后 ts 里面取出来有问题，得额外做一次编码转换，从 ansi 转成 utf8 。。估计。

**Reply by wangzhankun** [reply]
不太清楚。我跟onRun的代码做了对比，没发现有啥不一样的。不知道为啥onRun的就能正常获取到含有中文的路径。

**Reply by waruqi** [reply]
> 不太清楚。我跟onRun的代码做了对比，没发现有啥不一样的。不知道为啥onRun的就能正常获取到含有中文的路径。

onRun 没去取 target path 吧，只是直接回调 xmake run 命令，执行对应 target 了。。

而调试状态，targetpath 要直接丢给vscode的调试器，所以 xmake 后期不接管了。。targetpath编码依赖 调试器的编码需要

可以参考这段代码，判断如果是 iswin 平台后，用 encoding 对 target path 做下， utf8/gtk 的来回编码转换尝试，看看能否转成正常路径给调试器
https://github.com/xmake-io/xmake-vscode/blob/ec6667ad695e079febcb160f5bac607f04e131de/src/problem.ts#L47-L57

**Reply by wangzhankun** [reply]
好像没啥效果。gbk转utf8试过了，ansi转utf8也尝试了。但是转换之后仍然是乱码。
![image](https://user-images.githubusercontent.com/35724924/155837796-8a38e479-c671-4bba-b49e-6aaee94febd8.png)

**Reply by waruqi** [reply]
utf16 呢，显示乱码没关系，只要能 fs.existsSync 返回 true ，说明基本 ok 了

**Comment by wangzhankun** [comment]
我通过打印输出发现，返回的字符串就已经把中文部分的字符给替换成了65533.我在StackOverflow找到了[相关的描述](https://stackoverflow.com/questions/3526965/unicode-issue-with-an-html-title-question-mark-65533)。但是我对js不熟悉，不知道改怎么做调整。

![image](https://user-images.githubusercontent.com/35724924/155842189-2fc8d094-66e2-4545-911c-c2d20f1f3ba1.png)

![image](https://user-images.githubusercontent.com/35724924/155842162-a931f427-6b07-464e-87cf-66c873848d6d.png)

**Comment by wangzhankun** [comment]
在`process.ts`中的`iorunv`在转换为string之前进行判断是否为Windows并添加转换特别的转换语句可以解决问题。我建议添加一个专门的文件用来处理byte转为string的函数进行全局调用处理。
![image](https://user-images.githubusercontent.com/35724924/155842703-74d9b340-f2b3-4289-a601-c30dd72e3958.png)

**Comment by wangzhankun** [comment]
我提交了一个PR，xmake-io/xmake-vscode#85

**Reply by waruqi** [reply]
收到，非常感谢

**Reply by waruqi** [reply]
插件你先本地编译安装用着，发版本可能要等后面两天了，现在身边没环境提交

**Reply by waruqi** [reply]
我发布了，1.5.0 等几分钟后，更新下
